### DISCUSSION TITLE
Revisit discussion around precomputed errorbars

### DISCUSSION THREAD

**Discussion by JeppeKlitgaard** [mwaskom/seaborn#3347] [Q&A] [answered]
I have had a thorough look at the documentation and past issues on here, but feel sufficiently strongly about this feature that I thought I would present an argument for it again, though with a few additional points as well.

I want to start out with a bit of praise: I am really enjoying `seaborn` and it is a much saner tool to use than straight `matplotlib`. I've run into the same problem a few times now, however, which is the fact that `seaborn` natively supports calculating errorbars using bootstrapping, but the currently (as far as I can tell) isn't a way to specify precomputed errors.

Firstly, on a philosophical level I find it a bit too magical to have them computed for me, though this functionality can occasionally be nice and I am sure some people are really keen to do errorbars this way. Coming from physics, I often have dataframes with these precomputed using other techniques or need to do large bootstrappings which may be accelerated using custom code.

The previous argument against adding the option to specify precomputed errorbars has been that these can easily be implemented using `plt.errorbar`, but when working with a dataframe where you might be using a column as a categorial, this is non-trivial to implement.

Given there is already the code in place to render errorbars, would it not be fairly easy to implement `xerr` and `yerr` arguments that can specify columns specifying the error?

An example use-case:
```python
df = pd.DataFrame({
    "x": [1, 2, 3, 4, 5, 1, 2, 3, 4, 5],
    "y": [1, 2, 3, 4 ,5, 2, 4, 6, 8 ,10],
    "yerr": [0.5, 0.25, 0.5, 0.75, 0.5, 0.25, 0.5, 0.75, 0.5, 0.25],
    "cat": ["A"] * 5 + ["B"] * 5,
})
sns.lineplot(data=df, x="x", y="y", hue="cat")
# sns.lineplot(data=df, x="x", y="y", yerr="yerr", hue="cat")
```
![image](https://user-images.githubusercontent.com/1814611/234683720-f09c9343-158a-48e7-8f1d-f966c52b939a.png)

Ideally I would want to be able to use the bottom line.

Considerations:
- If `yerr` is defined ignore or warn if `errorbar` is set
- `xerr` is difficult with `lineplot`, but might for nice `scatterplot`
- `yerr` / `xerr` should follow similar logic as https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.errorbar.html

Apologies for reraising this issue, I understand completely if this is not something that seaborn wants to implement, but to me it seems like a really neat feature that _might_ be fairly easy to implement. I think it is a feature that quite a few people want, judging from:
- https://github.com/mwaskom/seaborn/issues/1887#issuecomment-1321401783
- https://github.com/mwaskom/seaborn/issues/331#issuecomment-562281923

**Comment by JeppeKlitgaard** [comment]
For reference, my current workaround is:
```python
ax = sns.lineplot(data=a, x="temp", y="abs_mag_dens_mean", hue="size", palette="tab10")
for i, (name, group) in enumerate(a.groupby("size")):
    x = group["temp"]
    y = group["abs_mag_dens_mean"]
    err = group["abs_mag_dens_sem"]
    color = ax.get_lines()[i].get_c()
    
    lower = y - err
    upper = y + err
    plt.fill_between(x, lower, upper, alpha=0.2, color=color)
```

**Comment by thuiop** [comment, marked as answer]
I guess you want one of these ?
![lineerrorbars](https://user-images.githubusercontent.com/1338337/234701339-eb394dae-3896-4057-8aba-95a296d07e42.png)
![linewithbands](https://user-images.githubusercontent.com/1338337/234701343-de770b52-1c1f-44d8-a245-e70890e3cfed.png)

Code is 
```
import seaborn.objects as so

df = pd.DataFrame({
    "x": [1, 2, 3, 4, 5, 1, 2, 3, 4, 5],
    "y": [1, 2, 3, 4 ,5, 2, 4, 6, 8 ,10],
    "yerr": [0.5, 0.25, 0.5, 0.75, 0.5, 0.25, 0.5, 0.75, 0.5, 0.25],
    "cat": ["A"] * 5 + ["B"] * 5,
})
df["ymin"] = df["y"] - df["yerr"]
df["ymax"] = df["y"] + df["yerr"]
p = (
    so.Plot(data=df, x="x", y="y", color="cat", ymin="ymin", ymax="ymax")
    .add(so.Line())
    .add(so.Range())
)
p.show()
```
(replace `Range` by `Band` to get the other one)

**Reply by JeppeKlitgaard** [reply]
This is fantastic! I hadn't gotten into the new `objects` API yet, but it seems really robust (albeit there are still some caveats, for example theming)

I still think the proposal would make sense for the old API, but perhaps this is being deprecated in the longer run?

**Reply by thuiop** [reply]
I believe the main dev is currently upgrading the old API to use some of the new objects and adding flexibility ; this might be something that will be introduced ? 
Anyway, do check out the new API ; it is not polished yet but I already find it pretty convenient.

**Reply by paw-lu** [reply]
Wanna say +1 and big thanks @thiop for this example. Was actually trying to do the exact same thing with `seaborn.objects` a few weeks ago but failed. Think it didn't click that I could use `ymax` and `Range` together like that.

**Comment by b-grimaud** [comment]
I just found this and I'm a bit surpised of how things worked out, I used seaborn 0.12.2 for while to generate barplots, and I would just pass a list of custom error values to `yerr=` and it worked perfectly. Updating to 0.13.0 now raises a `ValueError` : 

```
'yerr' (shape: (3,)) must be a scalar or a 1D or (2, n) array-like whose shape matches 'y' (shape: (1,))
ValueError: operands could not be broadcast together with remapped shapes [original->remapped]: (3,)  and requested shape (2,1)
```
